[MCKIN-8437] Restore /api/server/courses/{course_id}/completions view, using BlockCompletions.

jcdyer opened the pull request:
Implements an endpoint that was mistakenly removed, but required for group-project-v2.  The endpoint previously used CourseModuleCompletions as its backing data, but now uses BlockCompletions.  The response format includes all the fields the endpoint previously included, but now also includes the data under the new names used by the completion api.

**JIRA tickets**: [MCKIN-8437](https://edx-wiki.atlassian.net/browse/MCKIN-8437) (partial fix), [BB-166](https://tasks.opencraft.com/browse/BB-166).

**Discussions**: N/A
**Dependencies**: N/A

**Screenshots**: N/A

**Sandbox URL**: N/A
**Merge deadline**: None

**Testing instructions**: 

1. Pip install this branch into a copy of the solutions devstack development branch.
2. Enable the waffle switch in the admin: completion.enable_completion_tracking.
3. Log in, register for the demo course, and navigate through the first few pages (spend at least five seconds on each page, and scroll all the way to the bottom, to ensure that completions get submitted).
4. Send a request to http://lms.mcka.local/api/server/courses/course-v1:edX+DemoX+Demo_Course/completions, specifying either ?content_id=<block_key> or ?user_id=<user_id> (or both, or neither).  Verify that useful results are returned.  

**Author notes and concerns**:

N/A

**Reviewers**
- [ ] @lgp171188 

**Settings**
N/A

jcdyer commented on edx_solutions_api_integration/courses/views.py:
```diff
@@ -3,58 +3,47 @@
-from lms.lib.comment_client.utils import CommentClientMaintenanceError, CommentClientRequestError
-from lxml import etree
+from lms.lib.comment_client.utils import (CommentClientMaintenanceError,
+                                          CommentClientRequestError)
```
Oh.  That's weird.  I always use the four-indent style.  I wonder how I missed that here....

jcdyer replied:
More readable, *and* in line with the [edx python style guide](https://edx.readthedocs.io/projects/edx-developer-guide/en/latest/style_guides/python-guidelines.html#syntax-and-organization).  I never use the deeply hanging indent style.  I wonder how it got in here...

lgp171188 commented on edx_solutions_api_integration/courses/views.py:
````diff
@@ -1930,6 +1919,108 @@ def get(self, request, course_id):  # pylint: disable=W0613,W0221
+
+    Filters can also be applied
+    ```/api/courses/{course_id}/completions/?user_id={user_id}```
+    ```/api/courses/{course_id}/completions/?content_id={content_id}&stage={stage}```
````
@jcdyer, the `stage` query string parameter has to be removed since it is not supported after the recent changes.

jcdyer replied:
Updated docstring.

lgp171188 commented on edx_solutions_api_integration/courses/views.py:
```diff
@@ -1930,6 +1919,108 @@ def get(self, request, course_id):  # pylint: disable=W0613,W0221
+            return Response({'message': _('content_id is missing')}, status.HTTP_400_BAD_REQUEST)
+        if not user_id:
+            return Response({'message': _('user_id is missing')}, status.HTTP_400_BAD_REQUEST)
+        if not course_exists(request, request.user, course_id):
```
@jcdyer, wouldn't it be cleaner to use a serializer/form class to do these validations? It is just a suggestion and not strictly required.

jcdyer replied:
I gave a shot at rewriting this, but using forms changes a lot of details about how the data is processed, and it ended up obfuscating the code more than it helped.  I think especially given that this is deprecated code already (which I am restoring from previously being accidentally deleted), and since there are no other uses of Forms in this repo, I'd rather not mess with it.

lgp171188 replied:
@jcdyer :+1:

lgp171188 commented on edx_solutions_api_integration/courses/views.py:
```diff
@@ -3,58 +3,52 @@
+from lms.lib.comment_client.utils import (
+    CommentClientMaintenanceError,
+    CommentClientRequestError
+)        import warnings
```
@jcdyer, this line is missing a newline character before the `import` statement which has to be on a separate line.

lgp171188 commented on edx_solutions_api_integration/courses/views.py:
```diff
@@ -1930,6 +1922,107 @@ def get(self, request, course_id):  # pylint: disable=W0613,W0221
+        """
+        POST /api/courses/{course_id}/completions/
+        """
+        warnings.warn(DeprecationWarning, "BlockCompletions should not be created this way.  Use the suggested APIs instead")
```
@jcdyer, the above function call has arguments in the wrong order.

```suggestion
        warnings.warn("BlockCompletions should not be created this way.  Use the suggested APIs instead", DeprecationWarning)
```

jcdyer replied:
Yep.  I fixed that one, then fixed it again during the rebase, then it must have broken again later in the rebase.

lgp171188 commented on edx_solutions_api_integration/courses/views.py:
```diff
@@ -1930,6 +1922,107 @@ def get(self, request, course_id):  # pylint: disable=W0613,W0221
+            return Response({'message': _('content_id is invalid')}, status.HTTP_400_BAD_REQUEST)
+
+        completion, created = BlockCompletion.objects.submit_completion(
+            user_id=user_id,
```
@jcdyer, there is no 'user_id` keyword argument for the `submit_completion` method. There is a `user` keyword argument which accepts a `User` object and not a `user_id`.

jcdyer commented:
@lgp171188 I've fixed the problems and rebased against the latest master branch.  Please review again.  Note: If you include a split-mongo block key, (block-v1:...) you will need to escape the `+` plus-signs in the query parameter (but not, it seems, in the URL path).

lgp171188 commented:
@jcdyer, thanks for the update. I will review this and let you know if everything is okay.

lgp171188 commented on edx_solutions_api_integration/courses/views.py:
```diff
@@ -2015,6 +2023,110 @@ def get(self, request, course_id):  # pylint: disable=W0613,W0221
+        """
+        POST /api/courses/{course_id}/completions/
+        """
+        warnings.warn(
```
@jcdyer, `warnings` has to be imported. I think the import got removed in the refactor/rebase.

lgp171188 commented on edx_solutions_api_integration/courses/views.py:
```diff
@@ -2015,6 +2023,110 @@ def get(self, request, course_id):  # pylint: disable=W0613,W0221
+        )
+        content_id = request.data.get('content_id', None)
+        user_id = request.data.get('user_id', None)
+        stage = request.data.get('stage', None)
```
@jcdyer, this variable stage is unused.

jcdyer replied:
Removed.

lgp171188 commented on edx_solutions_api_integration/courses/views.py:
```diff
@@ -2015,6 +2023,110 @@ def get(self, request, course_id):  # pylint: disable=W0613,W0221
+        POST /api/courses/{course_id}/completions/
+        """
+        warnings.warn(
+            DeprecationWarning,
```
@jcdyer, the order of the arguments is reversed. I think it got removed in the refactor/rebase.

jcdyer replied:
Yep.  I fixed that one, then fixed it again during the rebase.  It must have gotten lost later on in the rebase. :(

lgp171188 commented on edx_solutions_api_integration/courses/views.py:
```diff
@@ -2015,6 +2023,110 @@ def get(self, request, course_id):  # pylint: disable=W0613,W0221
+                    "content_id": "324dfgd",
+                    "stage": None,
+                    "created": "2014-06-10T13:14:49.878Z",
+                    "modified": "2014-06-10T13:14:49.914Z"
```
@jcdyer, the response I got while testing this in my devstack looked something like

```json
{
"id": 18,
"user_id": 5,
"content_id": "block-v1:edX+DemoX+Demo_Course+type@poll+block@44cc060f582743899d70b850097cf029",
"course_id": "course-v1:edX+DemoX+Demo_Course",
"stage": null,
"created": "2018-10-26T18:22:59.416943Z",
"modified": "2018-10-26T18:22:59.417711Z",
"course_key": "course-v1:edX+DemoX+Demo_Course",
"block_key": "block-v1:edX+DemoX+Demo_Course+type@poll+block@44cc060f582743899d70b850097cf029",
"block_type": "poll",
"completion": 1,
"user": 5
}
```
so this example response has to be updated.

jcdyer replied:
Updated.

jcdyer commented:
I resolved your issues, added tests, and kicked off a test run here: https://github.com/edx-solutions/edx-platform/pull/1267,

lgp171188 commented on edx_solutions_api_integration/courses/tests.py:
```diff
@@ -3518,3 +3519,81 @@ def test_social_metrics_leader_list_with_user_position(self):
+
+    def test_completion_post_endpoint(self):
+        with override_switch("completion.enable_completion_tracking", active=True):
+            response = self.do_post(self.completion_uri, {
```
@jcdyer, the variable `response` is unused. So you can ignore the return value of the method call.
```suggestion
            self.do_post(self.completion_uri, {
```
